### PR TITLE
Create a "DataTemplate" class to allow people to output data-driven narratives

### PR DESCRIPTION
**Pull request by tra38**
I have not yet written any documentation yet for this feature, but I just want to know whether this feature might be useful enough to merge into master. I have written tests to illustrate how data is to be fed into a new "DataTemplate" class that will decide what Calyx Grammar to invoke.

I will disclose that my pull request does have an ulterior motive. There is currently no open source library that can perform  the same functionality as [Automated Insights' Wordsmith](https://automatedinsights.com), so I hope that these commits could turn Calyx into a possible competitor. Certainly you can perform what Wordsmith does by manually typing out if/else statements (as I did in ["The Atheists Who Believe In God"](https://github.com/tra38/The-Atheists-Who-Believe-In-God), my entry for NaNoGenMo 2015), but it seems more convenient if this process can be abstracted away.

### CODE REVIEW

**Comment by tra38** [comment]
I also implemented in functionality to allow the DataTemplate render raw data input as well, provided that the grammar uses "erb syntax" when referring to the data.

Again, no documentation yet, but I want to see if this pull request will be accepted before I start the documentation process.

**Comment by tra38** [comment]
I had some time, so I wrote some documentation. I hope you like it.

**Comment by maetl** [comment]
Hi, sorry I didn’t get a chance to review this sooner.

I’ll go through the code and post some feedback. I think there are some good, creative ideas here, although it’s slightly different to the vision I had for how template substitution would be implemented in the high level API.

I do think the feature you’re addressing here is useful—I’ve wanted something like this myself, so I understand the use case/need for it.

File: README.md [hunk 1]
```diff
@@ -172,8 +189,110 @@ end
...
+Calyx is open source and provided under the terms of the MIT license. Copyright (c) 2015 Mark Rickerby, (c) 2016 Tariq Ali
+
+See the `LICENSE` file [included with the project distribution](https://github.com/tra38/calyx/blob/master/LICENSE) for more information.
+
+## History
+In November 2015, Mark Rickerby created Calyx and used that gem to create [choose-your-own adventure gamebooks](https://github.com/dariusk/NaNoGenMo-2015/issues/189). He later on wrote a [blog post](http://maetl.net/notes/storyboard/gamebook-of-dungeon-tropes) explaining his thought process.
+
+In January 2016, Tariq Ali forked Calyx and started adding in new features to turn Calyx into a useful tool for generating data-driven narratives (robojournalism).
+
+##Contributions
```
Comment (maetl) [inline README.md]: Frankly, this goes a little bit further beyond my initial intentions for this repository.

While I welcome the feedback and am really happy you made something useful out of it, I think that adding contribution guidelines to the README is beyond the scope of what I’d expect to see in a single pull request.

However, I agree contribution guidelines are probably needed. I guess I never really expected to see any contributions to this library, and wasn’t prepared for them, as it was mostly written for my own purposes at the time.

File: lib/calyx.rb [hunk 2]
```diff
@@ -173,4 +176,42 @@ def generate
       registry[:start].evaluate
     end
   end
-end
+
```
Comment (maetl) [inline lib/calyx.rb]: This whole section is probably a good candidate for breaking out into a separate `data_template.rb` file.

File: lib/calyx.rb [hunk 3]
```diff
@@ -1,4 +1,7 @@
+require 'erb'
+
 module Calyx
+  #The Grammar class and the Production module was written by Mark Rickerby in 2015, and licensed under the MIT license.
```
Comment (maetl) [inline lib/calyx.rb]: I’d much prefer it if the classes had descriptive comments for `yard` doc compatibility. Authorship can be inferred from looking at Git history.

File: spec/calyx_spec.rb [hunk 4]
```diff
@@ -84,3 +86,103 @@ class StringFormatters < Calyx::Grammar
     expect(grammar.generate).to eq('Hello world')
   end
 end
+
```
Comment (maetl) [inline spec/calyx_spec.rb]: I really like how thorough and structured these tests are.

File: lib/calyx/version.rb [hunk 5]
```diff
@@ -1,3 +1,3 @@
 module Calyx
-  VERSION = '0.5.2'.freeze
+  VERSION = '0.6.2'.freeze
```
Comment (maetl) [inline lib/calyx/version.rb]: If it goes up to the next minor version, the patch number can be reset, so `0.6.0`.

**Comment by maetl** [comment]
@tra38 thanks so much for your ideas and code. Really interesting to see some of your ideas for building data-driven narratives (the “Big Picture Data Analysis (Mean, Median, Mode, etc.)” on your fork is a fantastic idea and could be hugely valuable).

It’s good to see a structured API for arranging multiple text generators into a coherent whole. This is definitely a practical issue I’ve found when using `Calyx::Grammar`. The pattern for composing multiple text generators together was never quite clear, but the basic idea was to nest generators inside the rules of other generators. This works okay, but could definitely be improved upon. Having a separate DSL for composing them might make sense.

A few specific comments about the API changes:

- The original design was based around the idea of a uniform pattern for text generators where a random seed would be passed into the constructor, then a `#generate` method would return the resulting string of text. This same pattern would work for Markov chains, etc, as well as template grammars.
- Having two separate styles of template delimiter in the one string format (`<%= %>` and `{}`) is a bit jarring to me. I previously considered the extension point for syntax being symmetric to the existing syntax (eg: `{{ref}}` or `{ref!}` or something similar). The Erb syntax does open up the scope for composing actual Erb templates though, which could lead to some interesting other consequences for generating HTML documents.
- Conditionals. Great idea. I wonder how easy it would be to support this in the existing `rule` format?
- Template variables could be integrated with the existing API by extending `#generate` to take a `context` hash mapping keys to values. I’m hoping this could be supported without needing to introduce any additional dependencies or new syntax.
- Rules could be memoized which would have a similar effect to context variables, but could be generated within the grammar itself. Right now, rules evaluate each time they’re called—a memoized rule would evaluate once, and then every additional reference to the rule symbol would return the same value. This would simplify some of the boilerplate around handling names, etc, consistently.

**Comment by maetl** [comment]
Some more notes about memoization.

Can already do this now at the global level by injecting single values into the rule config:

```ruby
rule :protagonist, ['Vladimir', 'Estragon'].sample
```

Alternative is to add explicit memoization support to the grammar API:

```ruby
memo :protagonist, 'Vladimir', 'Estragon'
```

Alternative is to support a special memoization substitution syntax:

```ruby
rule :protagonist_name, 'Vladimir', 'Estragon'
rule :protagonist, '{{protagonist_name}}'
```

I think an addition like this would solve a huge number of the templating issues without requiring a completely separate template API to be used over the top (of course, it’s always possible to use a template API anyway, and it would have particular utility as a way to model specific narrative structures—where the grammar API is too general).

**Comment by maetl** [comment]
Direct data substitutions could also be handled by a minor change to the existing API (using a very similar approach to what you’ve suggested, but passing the hash into the call to `#generate` rather than the constructor):

```ruby
class Text < Calyx::Grammar
  # Q: How to handle delimiters?
end

person = PersonRepository.find_first

text = Text.new(seed)
text.generate(name: person.first_name, age: person.age)
```

File: lib/calyx.rb (hunk 2 above)
Comment (tra38) [inline lib/calyx.rb]: That makes a lot of sense; I'll move this into its own file then.

**Comment by tra38** [comment]
Thanks for the comments! Here is what I think might be a working plan:

- Keep data-template as a way to order code, but find a way to allow users to inject data into a Grammar::Calyx class, possibly by using memoization or a change to the API. Ideally, there should be no change to syntax or new template delimiters.
- See if there is any way to implement conditionals within Grammar::Calyx.
- Clean up the code (move code into new files, remove comments crediting users, etc.)

**Comment by tra38** [comment]
Hey Maetl, I saw your roadmap. I'm fine with it so long as you keep on doing regular releases of RubyGems. I plan on starting development on a new branch and submitting brand new pull requests using that branch. I'm a little hazy on what you meant by a "block constructor" though, so I may be working on implementing some of the features we discussed here previously.

I plan on starting work on Saturday, Feb. 13th. Hopefully I will be able to stick to that promise.

I'm closing this pull request since this won't really be merged into master. I'll open up a new pull request with new code then.

**Comment by maetl** [comment]
Thanks. I’ll post some contribution guidelines shortly. Feel free to open up new issues on the repo in advance to get feedback before writing any code. Thinking about some of your ideas, I wonder if they might be better suited to a standalone library (`robo-data-template`?), particularly if you want to integrate ERB and an API focused on narrative templates at a higher level.

With regards to the releases, yes, I am keen to publish more regular improvements on RubyGems. I had written most of the code for `0.6` a few weeks ago—just hadn’t gotten around to testing it and folding it into the main codebase.

By the way, if you need to make a breaking change to a fork, or extend its functionality, you can always reference it from a fork on GitHub directly, so that you’re not dependent on the upstream release process for installing changes you need. This means you can get the exact commit of the library you want at any point, from any repo, without ever needing to upload anything to RubyGems. Bundler supports this natively, and the [specific_install gem](https://github.com/rdp/specific_install) adds this for command line RubyGems too.

See:

- http://bundler.io/git.html
- https://github.com/rdp/specific_install

**Comment by tra38** [comment]
Thanks for the tip about referencing from a fork. I'm sure to keep that in mind. I'm also thinking about releasing my ideas as a standalone library (though I have to think of a new name for the library then), but in the meanwhile, I'll think of some ideas of how to contribute to your project and will write some issues soon when I get the chance.

**Pull request closed without merging**